Sofa/Component/Constraint/Lagrangian/Solver/src/sofa/component/constraint/lagrangian/solver/ConstraintForceExporter.inl

```diff
@@ -0,0 +1,146 @@
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+
+#include <sofa/component/constraint/lagrangian/solver/ConstraintForceExporter.h>
+
+#include <sofa/core/visual/VisualParams.h>
+#include <sofa/simulation/AnimateEndEvent.h>
+
+namespace sofa::component::constraint::lagrangian::solver
+{
+
+template <class DataTypes>
+ConstraintForceExporter<DataTypes>::ConstraintForceExporter()
+    : Inherited()
+    , d_constraintForces(initData(&d_constraintForces, "constraintForces", "(output) Forces (or wrenches if rigids) computed from the mechanical state using the constraint solver."))
+    , d_draw(initData(&d_draw, true, "draw", "(debug) draw forces (as a line) for each position of the mechanical object."))
+    , d_drawForceScale(initData(&d_drawForceScale, 1.0, "drawForceScale", "(debug) Scale to apply on the force (draw as a line)."))
+    , l_constraintSolver(initLink("constraintSolver", "Link to the constraint solver."))
+{
+    this->f_listening.setValue(true);
+
+}
+
+template <class DataTypes>
+void ConstraintForceExporter<DataTypes>::init()
+{
+    if (!l_constraintSolver)
+    {
+        l_constraintSolver.set(this->getContext()->template get<ConstraintSolverImpl>());
+    }
+
+    if (!l_constraintSolver)
+    {
+        msg_error() << "No constraint solver found in the current context, whereas it is required. This component retrieves the matrix from a linear solver.";
+        this->d_componentState.setValue(sofa::core::objectmodel::ComponentState::Invalid);
+        return;
+    }
+
+    // No need to check if the link to the mechanical state is correct
+    // as it has been covered by canCreate()
+    // Inherited::init();
+
+    this->d_componentState.setValue(sofa::core::objectmodel::ComponentState::Valid);
+}
+
+template <class DataTypes>
+void ConstraintForceExporter<DataTypes>::handleEvent(core::objectmodel::Event* event)
+{
+    if (this->d_componentState.getValue() != sofa::core::objectmodel::ComponentState::Valid)
+    {
+        return;
+    }
+
+    Inherited::handleEvent(event);
+
+    if (simulation::AnimateEndEvent::checkEventType(event))
+    {
+        computeForce();
+    }
+}
+
+template <class DataTypes>
+void ConstraintForceExporter<DataTypes>::draw(const sofa::core::visual::VisualParams* vparams)
+{
+    if (!d_draw.getValue() || this->d_componentState.getValue() != sofa::core::objectmodel::ComponentState::Valid)
+    {
+        return;
+    }
+
+    auto constraintForces = sofa::helper::getReadAccessor(d_constraintForces);
+    const auto& positions = this->getMState()->read(sofa::core::ConstVecCoordId::position())->getValue();
+
+    assert(constraintForces.size() != positions.size());
+
+    const float lineSize = 2.0f;
+    const double forceScale = d_drawForceScale.getValue();
+
+    vparams->drawTool()->saveLastState();
```

Comment:
isn't there a RAII for that ?

Reply:
If not, that's an excellent idea

Reply:
I dont think so, and I second that this is a good idea

Reply:
https://github.com/sofa-framework/sofa/pull/3338